src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -149,15 +177,340 @@
+    invoice.withFolioInvoiceNo(invoiceFromStorage.getFolioInvoiceNo());
+  }
+
+  private CompletionStage<Void> handleInvoiceStatusTransition(Invoice invoice, Invoice invoiceFromStorage) {
+    if (isTransitionToApproved(invoiceFromStorage, invoice)) {
+      return approveInvoice(invoice);
+    } else if (isTransitionToPaid(invoiceFromStorage, invoice)) {
+      return payInvoice(invoice);
+    }
+    return VertxCompletableFuture.completedFuture(null);
+  }
+
+  private boolean isTransitionToApproved(Invoice invoiceFromStorage, Invoice invoice) {
+    return (invoiceFromStorage.getStatus() == Invoice.Status.REVIEWED || invoiceFromStorage.getStatus() == Invoice.Status.OPEN) && invoice.getStatus() == Invoice.Status.APPROVED;
+  }
+
+  /**
+   * Handles transition of given invoice to {@link Invoice.Status#APPROVED} status.
+   * Transition only triggers if the current {@link Invoice.Status} is {@link Invoice.Status#REVIEWED}
+   * and exist at least one {@link InvoiceLine} associated with this invoice
+   *
+   * @param invoice {@link Invoice}to be approved
+   * @return CompletableFuture that indicates when transition is completed
+   */
+  private CompletableFuture<Void> approveInvoice(Invoice invoice) {
+    CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+    getInvoiceLinesWithTotals(invoice)
+      .thenApply(invoiceLines -> {
+        verifyInvoiceLineNotEmpty(invoiceLines);
+        return invoiceLines;
+      })
+      .thenCombine(prepareVoucher(invoice), (invoiceLines, voucher) -> {
+        invoice.setVoucherNumber(voucher.getVoucherNumber());
+        return handleVoucherWithLines(invoiceLines, voucher)
+          .thenAccept(future::complete)
+          .exceptionally(t -> {
+            future.completeExceptionally(t);
+            return null;
+          });
+      })
+    .exceptionally(t -> {
+      future.completeExceptionally(t);
+      return null;
+    });
+    return future;
+  }
+
+  private void verifyInvoiceLineNotEmpty(List<InvoiceLine> invoiceLines) {
+    if (invoiceLines.isEmpty()) {
+      throw new HttpException(500, NO_INVOICE_LINES_ERROR_MSG);
+    }
+  }
+
+  /**
+   * Prepares a new voucher or updates existing one for further processing
+   *
+   * @param invoice {@link Invoice} to be approved on the basis of which the voucher is prepared
+   * @return completable future with {@link Voucher} on success
+   */
+  private CompletableFuture<Voucher> prepareVoucher(Invoice invoice) {
+    return findExistingVoucher(invoice.getId())
+      .thenCompose(vouchers -> {
+        if (isNotEmpty(vouchers)) {
+          Voucher voucher = vouchers.get(0);
+          return updateExistingVoucherState(invoice, voucher);
         }
+        return buildNewVoucher(invoice);
       });
   }
 
+  private CompletableFuture<List<Voucher>> findExistingVoucher(String invoiceId) {
+    return voucherHelper.getVouchers(1, 0, String.format(QUERY_BY_INVOICE_ID, invoiceId))
```

Comment:
There is `getVoucherByInvoiceId(String invoiceId)` method in `VoucherHelper` which might be used instead

Reply:
Removed